src/mlpack/core/data/binarize.hpp

```diff
@@ -79,13 +80,60 @@ void Binarize(const arma::Mat<T>& input,
               const double threshold,
               const size_t dimension)
 {
+  // stopping invalid memory access and crashes
+  if (dimension >= input.n_rows)
+  {
+    throw std::invalid_argument("Invalid value for dimension");
+  }
   output = input;
 
   #pragma omp parallel for
   for (omp_size_t i = 0; i < (omp_size_t) input.n_cols; ++i)
     output(dimension, i) = input(dimension, i) > threshold;
 }
-
+/**
+ * Given an input dataset and threshold, set values greater than threshold to
+ * 1 and values less than or equal to the threshold to 0. This overload takes
+ * a vector of dimension and applys the changes to the given vector of dimension.
+ *
+ * @code
+ * arma::Mat<double> input = loadData();
+ * arma::Mat<double> output;
+ * double threshold = 0.5;
+ * vector<int> dimension = {1,2};
+ *
+ * // Binarize the second and third dimension. All positive values in the second
+ * // and thirds dimensionwill be set to 1 and the values less than or equal
+ * // to 0 will become 0.
+ * Binarize<double>(input, output, threshold, dimension);
+ * @endcode
+ *
+ * @param input Input matrix to Binarize.
+ * @param output Matrix you want to save binarized data into.
+ * @param threshold Threshold can by any number.
+ * @param vector of dimension Feature to apply the Binarize function.
+ */
+template<typename T>
+void Binarize(const arma::Mat<T>& input,
+              arma::Mat<T>& output,
+              const double threshold,
+              std::vector<int>row)
+{
+  output = input;
```

Comment:
I'm wondering if this is something that is used in practice, in most cases I want to binarize the complete matrix, and not sure I can encounter a case where I wanted to binarize a single row, but maybe this is something that is used?

Would be interesting to get some input from the community on this one.

Reply:
I don't know of a case where I'd like to binarize multiple rows. For a single row I understand it might get useful for setting the test labels for binary classification, but that's a guess in itself. I'd be +0 for this change.

Reply:
@zoq @ShikharJ I agree that, it is not often that you have to binarize multiple rows, but it is also used in discretization of continuous feature values.

You may be right that this overloaded function may not be used by many , but i have encountered occassion where i have to binarize multiple columns and not row depending on different threshold so may be we could over load the function to take multiple threshold to binarize, something like

`void Binarize(const arma::Mat<T>& input,
              arma::Mat<T>& output,
              vector<double> threshold,
              vector<int>row)`

You could find an example [here](https://www.geeksforgeeks.org/sklearn-binarizer-in-python/) , an article in geeksforgeeks, where they have binarized multiple column using different threshold

Would like to here your take on this too.

Also there is [package](https://rdrr.io/github/LudvigOlsen/LRO.utilities/man/binarizer.html) in R,that would binarize multiple columns using same threshold

i would also like to have your input on : the existing overloaded function
 
`void Binarize(const arma::Mat<T>& input,
              arma::Mat<T>& output,
              const double threshold,
              const size_t dimension)`

which Binarizes a row, I also don't know why would anyone do that, I would surely binarize a column and not a row, as @ShikharJ suggested for  setting the test labels for binary classification, Since test labels would be along the column and not along the row.Binarizing a row would make the features go off track because every row would have some features consistency, except the row we binarized.

Maybe we could,change the function to binarize a particular column and not row.

Reply:
@jeffin143 In mlpack, each column is a data point. So test labels should be along the row.

Reply:
> @jeffin143 In mlpack, each column is a data point. So test labels should be along the row.

oh i see.

Reply:
I can definitely see the utility of binarizing the entire matrix, and that would probably be the "typical" case.  But I think it might still be useful to have a function for just one dimension, since you might want to set a different threshold for each dimension that you are binarizing.  I wasn't able to think of any common cases where you would want to binarize some dimensions but not all of them---but I suppose it's possible someone might want to do that (I am just not totally sure why).

Reply:
@jeffin143 Thanks for the links, personally I'm inclined to discard the feature; if somebody depends on it, a simple for loop around the existing:

```c++
void Binarize(const arma::Mat<T>& input,
              arma::Mat<T>& output,
              const double threshold,
              const size_t dimension)
```

function will do the same. Don't have a strong opinion about this, in case anybody likes to get this merged.

Reply:
@zoq , I agree with you. Closing the PR . Thanks for your input.

Reply:
Even if we ended up not merging this in, it was still a good contribution.